### PR TITLE
[possibly-used-before-assignment] Avoid FP for typing.NoReturn

### LINKED ISSUE #9674: False positive: `possibly-used-before-assignment` and `pytest.skip` (and other `NoReturn` functions)
### Bug description

```python
import pytest
import sys

def test_something():
    if sys.platform == "linux":
        cmd = "ls"
    elif sys.platform == "win32":
        cmd = "dir"
    else:
        pytest.skip("only runs on Linux/Windows")

    print(cmd)
```

### Configuration

_No response_

### Command used

```shell
pylint -E test_x.py
```

### Pylint output

```python
************* Module test_x
test_x.py:12:10: E0606: Possibly using variable 'cmd' before assignment (possibly-used-before-assignment)
```

### Expected behavior

No error: The `pytest` module has a couple of functions that never return:

- `pytest.exit(...)`
- `pytest.skip(...)`
- `pytest.fail(...)`
- `pytest.xfail(...)`

Related:

- #9627 
- #9643 

---

Instead of maintaining a list of functions like this, could pylint maybe instead check whether they have a [`NoReturn` return type](https://docs.python.org/3/library/typing.html#typing.NoReturn), if type information is available? That would cover `assert_never()`, all the pytest functions, possibly `sys.exit()` and friends (if stdlib typing information is available via typeshed), and lots and lots of other third-party cases like this.

### Pylint version

```shell
pylint 3.2.1
astroid 3.2.2
Python 3.12.3 (main, Apr 23 2024, 09:16:07) [GCC 13.2.1 20240417]
```

### OS / Environment

Archlinux

### Additional dependencies

_No response_

### PR DESCRIPTION
**Pull request by jacobtylerwalls**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description
Closes #9674 for the general case, but not necessarily for pytest.skip due to the following:

```py
from astroid import extract_node
n = extract_node("""
... import pytest
... pytest.skip('foo')""")
>>> n.func.inferred()
[Uninferable]
```

### CODE REVIEW

File: pylint/checkers/utils.py [hunk 1]
```diff
@@ -2165,6 +2169,24 @@ def is_terminating_func(node: nodes.Call) -> bool:
...
+            if (
+                isinstance(inferred, nodes.FunctionDef)
+                and isinstance(inferred.returns, nodes.Name)
+                and (inferred_func := safe_infer(inferred.returns))
+                and hasattr(inferred_func, "qname")
+                and inferred_func.qname()
+                in (
+                    *TYPING_NORETURN,
+                    # In Python 3.7 - 3.8, NoReturn is alias of '_SpecialForm'
+                    "typing._SpecialForm",
```
Comment (jacobtylerwalls) [inline pylint/checkers/utils.py:2186]: We'll be dropping support for 3.8 in the next minor release, so this inelegance doesn't bother me ATM.

**Comment by Pierre-Sassoulas** [comment]
I'm on the fence about this one. On the one hand, it's going to prevent false positives. And we have false positives for code that is too dynamic for astroid to understand. But on the other hand, one of the value proposition of pylint is not trusting the typing. If we start trusting the typing then we're going to raise the same class of error that type checker already raise, and they do that better than us and faster than us. Also, if we don't do it generically we're going to have some message that trust the typing and some that don't and how are we drawing the line when someone ask for typing use making the point that "possibly-user-before-assignment" does it for "NoReturn" already ? Then again I don't see massive improvements happening in astroid's inference any time soon.

**Comment by jacobtylerwalls** [comment]
I hear that. I think NoReturn is very special -- it's basically assert_never, which is basically a more expressive pylint disable.

Allowing someone to use a stdlib feature that works well with type checkers instead of a pylint disable is a pretty isolated case compared to _always trusting typing_.

**Comment by Pierre-Sassoulas** [comment]
I like the idea if seeing some type hint as implicit disables. A passing thought : Taking other tool's noqas into account is somewhat similar and would also make pylint more useful.

**Review by Pierre-Sassoulas — DISMISSED** [review]
I like the fact that the genericity if the code permits to also makes 'unreachable' better. It's done properly because it's easy to see that kind of helper reuse as a dirty source of unintented side effects.

**Comment by Pierre-Sassoulas** [comment]
@DanielNoord do you want to chime in :) ?

**Review by DanielNoord — COMMENTED** [review]
I like it! However, I would like to document that this is a special case and not something we do for other checks. Then we have something we can point to when more requests start coming in.

**Comment by jacobtylerwalls** [comment]
Other checks or other type annotation concepts? I've made change in the general util, I think we probably should simplify the code for other checks by using the common utils more, so I don't know there's much more to document.

**Comment by DanielNoord** [comment]
> Other checks or other type annotation concepts?

Type annotation concepts. I meant in the `details.rst` of the message: explain that we made an exception for this message to take this into account but have no current plans to start taking type annotation into account.

**Review by DanielNoord — DISMISSED** [review]

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM. (The last commit's CI need to run so I guess we should remove [ci skip])

**Comment by jacobtylerwalls** [comment]
Sounds good, I forgot that this repo doesn't really allow that. I'll squash and push.

**Review by Pierre-Sassoulas — APPROVED** [review]
Wonderful I can just squash with the original author intended commit message ™️ 

Do you agree with a release just after the backport ?

**Pull request merged**